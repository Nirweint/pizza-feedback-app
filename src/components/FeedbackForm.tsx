import React from "react";
import {useForm, Controller, SubmitHandler} from "react-hook-form";
import {
    Paper,
    Typography,
    Box,
    Rating,
    TextField,
    FormControl,
    FormGroup, Button
} from "@mui/material";
import {getLocalStorageState, setLocalStorageState} from "../localStorage";
import {FeedbackType} from "../types";


type FeedbackFormType = Omit<FeedbackType, 'name'>

type FeedbackFormPropsType = {
    name: string
    setOpenModal: (value: boolean) => void
}

export const FeedbackForm = ({name, setOpenModal}: FeedbackFormPropsType) => {
    const feedbacksFromLocalStorage = getLocalStorageState<FeedbackType[]>('feedback', [])

    const {
        handleSubmit,
        control,
        formState: {errors, touchedFields}
    } = useForm<FeedbackFormType>({mode: 'onTouched'});
    const onSubmit: SubmitHandler<FeedbackFormType> = data => {
        const dataToSend: FeedbackType[] = [...feedbacksFromLocalStorage, {...data, name}]
        setLocalStorageState('feedback', dataToSend)
        setOpenModal(false)
    };

    const handleCloseModalClick = () => {
        setOpenModal(false)
    }

    const touched = Object.keys(touchedFields)
    const hasErrors = Object.keys(errors)

    return (
        <Paper elevation={3}>
            <Box p={2} sx={{display: 'flex', justifyContent: 'center'}}>

                <FormControl>
                    <form onSubmit={handleSubmit(onSubmit)}>
                        <Typography>Name</Typography>
                        <Typography variant={"h4"}>{name}</Typography>
                        <FormGroup>
                            <Box sx={{margin: '10px 0'}}>
                                <Controller
                                    name="rating"
                                    control={control}
                                    defaultValue={3}
                                    rules={{required: true}}
                                    render={({field: {value, onChange, name, ref}}) => {
                                        return <Rating
                                            value={+value}
                                            onChange={onChange}
                                            name={name}
                                            ref={ref}
                                        />
                                    }}
                                />
                            </Box>

                            <Box>
                                <Controller
                                    name="phone"
                                    control={control}
                                    defaultValue={''}
                                    // не знаю что это за номер длиной в максимум 10 символов, но вот
                                    rules={{
                                        required: "Please enter your phone number",
                                        minLength: {
                                            value: 3,
                                            message: 'Min length 3 chars',
                                        },
                                        maxLength: {
                                            value: 10,
                                            message: 'Max length 10 chars',
                                        },
                                        pattern: {
                                            value: /^[0-9/+/ /(/)]+$/,
                                            message: "Please enter a valid phone number"
                                        }
                                    }}
                                    render={({field}) => {
                                        return <TextField
                                            required
                                            margin="normal"
                                            {...field}
                                            label="phone"
                                            helperText={errors.phone?.message}
                                            error={!!errors.phone}
                                        />
                                    }}
                                />
                            </Box>

                            <Box sx={{margin: '10px 0'}}>
                                <Controller
                                    name="comment"
                                    control={control}
                                    defaultValue={''}
                                    rules={{required: true,}}
                                    render={({field}) => <TextField
                                        {...field}
                                        required
                                        label="comment"
                                        multiline rows={4}/>}
                                />
                            </Box>

                            <Box>
                                {touched.length >= 2 && hasErrors.length === 0 ?
                                    <Button
                                        type="submit"
                                        color={"success"}
                                        variant="outlined">
                                        SAVE
                                    </Button>
                                    :
                                    <Button
                                        onClick={handleCloseModalClick}
                                        color={"error"}
                                        variant="outlined">
                                        CANCEL
                                    </Button>
                                }
                            </Box>
                        </FormGroup>
                    </form>
                </FormControl>

            </Box>
        </Paper>
    );
}