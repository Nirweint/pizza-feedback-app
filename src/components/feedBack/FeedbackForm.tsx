import React, {useState} from "react";
import {useForm, Controller, SubmitHandler} from "react-hook-form";
import {
  Paper,
  Typography,
  Box,
  Rating,
  TextField,
  FormControl,
  FormGroup,
  Button
} from "@mui/material";
import {FeedbackType} from "../../types";
import {getLocalStorageState, setLocalStorageState} from "../../localStorage";
import {v1} from 'uuid';
import {useDispatch} from "react-redux";
import {setFeedbackAC} from "../../store/reducers/feedback";

type FeedbackFormType = Omit<FeedbackType, "name">;

type fieldsArrayType = { id: string }[];

type FeedbackFormPropsType = {
  name: string;
  setOpenModal: (value: boolean) => void;
};

export const FeedbackForm = ({name, setOpenModal}: FeedbackFormPropsType) => {

  const dispatch = useDispatch()
  const [fields, setFields] = useState<fieldsArrayType>([]);

  const feedbacksFromLocalStorage = getLocalStorageState<FeedbackType[]>(
    "feedback",
    []
  );

  const {
    handleSubmit,
    control,
    formState: {errors, touchedFields}
  } = useForm<FeedbackFormType>({mode: "onTouched"});

  const onSubmit: SubmitHandler<FeedbackFormType> = (data) => {
    const dataToSend: FeedbackType[] = [
      ...feedbacksFromLocalStorage,
      {...data, name}
    ];
    setLocalStorageState("feedback", dataToSend);
    dispatch(setFeedbackAC(dataToSend))
    setOpenModal(false);
  };

  const handleCloseModalClick = () => {
    setOpenModal(false);
  };

  const handleCreateNewFieldClick = () => {
    const newId = v1()

    setFields([{id: newId}, ...fields])
  }

  const touched = Object.keys(touchedFields);
  const hasErrors = Object.keys(errors);

  return (
    <div>
      <Paper elevation={3} sx={{
        overflowY: 'scroll',
        maxHeight: '600px',
        "::-webkit-scrollbar": {display: "none", scrollbarWidth: 'none'}
      }}>
        <Box p={2} sx={{display: "flex", justifyContent: "center", width: 'auto'}}>
          <FormControl>

            {fields.map(({id}) => {
              return (
                <Box sx={{margin: "10px 0"}} key={id} id={id} data-testid='empty-input'>
                  <TextField margin="normal" type='text' />
                </Box>
              )
            })}

            <form onSubmit={handleSubmit(onSubmit)}>
              <Typography>Name</Typography>
              <Typography variant={"h4"}>{name}</Typography>
              <FormGroup>
                <Box sx={{margin: "10px 0"}}>
                  <Controller
                    name="rating"
                    control={control}
                    defaultValue={3}
                    rules={{required: true}}
                    render={({field: {value, onChange, name, ref}}) => {
                      return (
                        <Rating
                          value={+value}
                          onChange={onChange}
                          name={name}
                          ref={ref}
                        />
                      );
                    }}
                  />
                </Box>

                <Box>
                  <Controller
                    name="phone"
                    control={control}
                    defaultValue={""}
                    rules={{
                      required: "Please enter your phone number",
                      minLength: {
                        value: 3,
                        message: "Min length 3 chars"
                      },
                      maxLength: {
                        value: 10,
                        message: "Max length 10 chars"
                      },
                      pattern: {
                        value: /^[0-9/+/ /(/)]+$/,
                        message: "Please enter a valid phone number"
                      }
                    }}
                    render={({field}) => {
                      return (
                        <TextField
                          required
                          margin="normal"
                          {...field}
                          label="phone"
                          helperText={errors.phone?.message}
                          error={!!errors.phone}
                        />
                      );
                    }}
                  />
                </Box>

                <Box sx={{margin: "10px 0"}}>
                  <Controller
                    name="comment"
                    control={control}
                    defaultValue={""}
                    rules={{required: true}}
                    render={({field}) => (
                      <TextField
                        {...field}
                        required
                        label="comment"
                        multiline
                        rows={4}
                      />
                    )}
                  />
                </Box>

                <Box sx={{margin: "10px 0"}}>
                  {touched.length >= 2 && hasErrors.length === 0 ? (
                    <Button type="submit" color={"success"} variant="outlined">
                      SAVE
                    </Button>
                  ) : (
                    <Button
                      onClick={handleCloseModalClick}
                      color="error"
                      variant="outlined"
                    >
                      CANCEL
                    </Button>
                  )}
                  <Button color="success" variant="outlined" sx={{marginLeft: '5px'}}
                          onClick={handleCreateNewFieldClick}>
                    ADD NEW FIELD
                  </Button>
                </Box>
              </FormGroup>
            </form>
          </FormControl>
        </Box>
      </Paper>
    </div>
  );
};
