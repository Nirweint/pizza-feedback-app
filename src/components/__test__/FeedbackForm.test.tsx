import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import { FeedbackForm } from "../FeedbackForm";

describe('FeedbackForm', () => {

  test('renders add new field button', () => {
    render(<FeedbackForm name={"Alex"} setOpenModal={() => {}}/>)

    expect(screen.getByText(/ADD NEW FIELD/i)).toBeInTheDocument();
  });

  test('renders correct button', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)

    const cancelButton = screen.getByText(/CANCEL/i);

    expect(cancelButton).toBeInTheDocument();
    expect(cancelButton).toMatchSnapshot();
    expect(cancelButton).toHaveAttribute('type', 'button');
  })

  test('renders correct user name', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)
    expect(screen.getByText(/Alex/i)).toBeInTheDocument();

    render(<FeedbackForm name={'Andrey V'} setOpenModal={() => {}}/>)
    expect(screen.getByText(/Andrey V/i)).toBeInTheDocument();
  })

  test('renders all inputs', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)

    const phoneInput = screen.getByLabelText(/phone/i);
    const commentInput = screen.getByLabelText(/comment/i);
    const rating = screen.getByDisplayValue(/3/i);

    expect(phoneInput).toBeInTheDocument();
    expect(commentInput).toBeInTheDocument();
    expect(rating).toBeInTheDocument();
  })

  test('correct default value in rating input', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)

    const rating = screen.getByDisplayValue(/3/i);

    expect(rating).toBeInTheDocument();
    expect(rating).toHaveAttribute('value', '3')
    expect(rating).toMatchSnapshot();
  })

  test('adding inputs on addNewField click', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)
    const addNewFieldButton = screen.getByText(/ADD NEW FIELD/i);

    userEvent.click(addNewFieldButton)
    userEvent.click(addNewFieldButton)

    expect(addNewFieldButton).toBeInTheDocument();
    expect(screen.getAllByTestId('empty-input').length).toBe(2);

    userEvent.click(addNewFieldButton)

    expect(screen.getAllByTestId('empty-input').length).toBe(3);

  })

  test('new inputs must be on top of the form', () => {
    render(<FeedbackForm name={'Alex'} setOpenModal={() => {}}/>)
    const addNewFieldButton = screen.getByText(/ADD NEW FIELD/i);

    userEvent.click(addNewFieldButton)
    expect(addNewFieldButton).toBeInTheDocument();
    expect(screen.getAllByTestId('empty-input').length).toBe(1);
    const firstInputId = screen.getByTestId('empty-input').getAttribute('id');

    userEvent.click(addNewFieldButton);

    expect(screen.getAllByTestId('empty-input').length).toBe(2);
    expect(screen.getAllByTestId('empty-input')[1].getAttribute('id')).toBe(firstInputId);
  })

})