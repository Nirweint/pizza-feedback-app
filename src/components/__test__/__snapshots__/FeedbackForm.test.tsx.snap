// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackForm correct default value in rating input 1`] = `
<input
  checked=""
  class="MuiRating-visuallyHidden"
  id="mui-38"
  name="rating"
  type="radio"
  value="3"
/>
`;

exports[`FeedbackForm renders correct button 1`] = `
<button
  class="MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-180lf2n-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  type="button"
>
  CANCEL
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
